### PR TITLE
Set action by default in stimulus integration

### PR DESCRIPTION
**Pull request by amiel**
## what

This updates the default error handler for the stimulus integration to call `setAction` using the name of the stimulus controller where the error was thrown.

I also included some tests for the stimulus integration, since there were none yet.

## why

We've been using `@appsignal/stimulus` for our frontend exception tracking for a while and it's working great, except that all errors end up being grouped together which caused some confusion.

Screenshot 2024-12-17 at 15 01 03

We've already started using this same idea at [Bento](https://github.com/bentohq) by writing our own custom error handler. This results in grouping errors by controller, which is similar to how errors are grouped by default for Ruby on Rails backend web requests.

Screenshot 2024-12-17 at 15 04 17

## notes

Apologies if this is the wrong direction. I didn't create an issue before implementing this pull-request based on one of the tasks from #613, which seems to have broad approval:

> Report the action in which an error takes place (for example, for a React component, attempt to report the component name)

This is also similar to [how it's done for vue](https://github.com/appsignal/appsignal-javascript/blob/ef1265730fe1eef2dbd52bc71027d3ae90a744b6/packages/vue/src/index.ts#L14), which is in place even after #41.

### safety

Based on the code that calls the error handler, it looks like there should always be details with an identifier. This code has been in place since stimulus v1.0.0-beta1.

https://github.com/hotwired/stimulus/blob/1c425edb7d71b05ee85e76becdd1620988042de2/packages/%40stimulus/core/src/context.ts#L117-L119

If there's a strong opinion, I'd be happy to make this more defensive, but I don't think it's necessary.

### CODE REVIEW

File: packages/stimulus/src/index.ts [hunk 1]
```diff
@@ -3,9 +3,12 @@ import type { JSClient, JSSpan } from "@appsignal/types"
 export function installErrorHandler(appsignal: JSClient, application: any) {
   const prevHandler = application.handleError
 
-  application.handleError = function (error: Error, message: string) {
```
Comment (amiel) [inline packages/stimulus/src/index.ts:6]: This space was removed by eslint auto-formatting, so I _assume_ it is consistent with the project style guide. Let me know if I need to do something different here.

File: packages/stimulus/src/index.ts [hunk 2]
```diff
@@ -3,9 +3,12 @@ import type { JSClient, JSSpan } from "@appsignal/types"
...
+  application.handleError = function(error: Error, message: string, detail: { identifier: string }) {
```
Comment (amiel) [inline packages/stimulus/src/index.ts:6]: Stimulus types simply define the [detail as `object`](https://github.com/hotwired/stimulus/blob/101a060f961e9240b55b29e25fdfe82c79b9c5b0/src/core/error_handler.ts#L2), so I didn't bother trying to pull in the real type here. Open to suggestions...

**Comment by unflxw** [comment]
Hi @amiel, thank you so much for contributing this!

File: packages/stimulus/src/index.ts (hunk 2 above)
Comment (unflxw) [reply packages/stimulus/src/index.ts]: I think not pulling in the real type is fine.

Out of an abundance of caution, we could account for `detail` potentially not including `identifier` by typing it as an optional attribute, and only setting the action if it is present. We do try our best to avoid being the reason why someone's application crashes in production.

That said, given that the code that invokes `handleError` has remained unchanged this long, and that there don't seem to be any circumstances in which a controller does not have an identifier, I'm not really concerned about it either way.

**Review by unflxw — APPROVED** [review]

File: packages/stimulus/src/index.ts (hunk 2 above)
Comment (amiel) [reply packages/stimulus/src/index.ts]: > Out of an abundance of caution, we could account for `detail` potentially not including `identifier`

I'm going to actually go ahead and do this. Having looked at the code, I feel pretty confident it'll always be present, but it also seems reasonable for error handlers to be extra defensive, and this would also be some future-proofing insurance.

File: packages/stimulus/src/index.ts [hunk 3]
```diff
@@ -3,9 +3,16 @@ import type { JSClient, JSSpan } from "@appsignal/types"
...
-  application.handleError = function (error: Error, message: string) {
+  application.handleError = function (
+    error: Error,
+    message: string,
+    detail: { identifier: string }
+  ) {
     const span = appsignal.createSpan((span: JSSpan) =>
-      span.setTags({ framework: "Stimulus", message }).setError(error)
+      span
+        .setAction(`${detail.identifier}-controller`)
```
[the remark this answers is not in the document]
Comment (amiel) [reply packages/stimulus/src/index.ts:13]: > but my guess is that it is trying to make the action "look like" [the file location that this identifier is derived from](https://stimulus.hotwired.dev/reference/controllers#identifiers)?

That was exactly my intention.

> Do you think it be useful to attempt to use detail.controller.constructor.name as the action name?

This makes a lot of sense, but based on common usage, I think we should expect it to be an anonymous class.

It would be possible to use the class name and fall back to the identifier with something like...

```js
let name = detail.controller.constructor.name
if (name === "extended") { name = detail.identifier }
```

I suggest keeping it simpler, but let me know if you prefer this.

---

As a proposal, I'm going to change this to just use the identifier. Although it may be less obvious what it refers to, stimulus developers will be comfortable with it from defining `data-controller="identifier"`. This feels like the safest and least-surprising approach.

That said, I'm happy to continue discussing the most appropriate default for this.

[the remark this answers is not in the document]
Comment (unflxw) [reply packages/stimulus/src/index.ts:13]: > I think we should expect it to be an anonymous class. It would be possible to use the class name and fall back to the identifier [...] I suggest keeping it simpler

Agreed. I also think it would be confusing if some controllers ended up showing the class name while others used the identifier.

> As a proposal, I'm going to change this to just use the identifier. Although it may be less obvious what it refers to, stimulus developers will be comfortable with it from defining data-controller="identifier". This feels like the safest and least-surprising approach.

That sounds reasonable to me as well! Go ahead 💯👍

File: packages/stimulus/src/index.ts [hunk 4]
```diff
@@ -3,14 +3,14 @@ import type { JSClient, JSSpan } from "@appsignal/types"
...
+  application.handleError = function(
     error: Error,
     message: string,
-    detail: { identifier: string }
+    detail: { identifier?: string }
   ) {
     const span = appsignal.createSpan((span: JSSpan) =>
       span
-        .setAction(`${detail.identifier}-controller`)
+        .setAction(detail?.identifier || "[unknown Stimulus controller]")
```
Comment (amiel) [inline packages/stimulus/src/index.ts:13]: This default string is patterned off of the [vue integration](https://github.com/appsignal/appsignal-javascript/blob/ef1265730fe1eef2dbd52bc71027d3ae90a744b6/packages/vue/src/index.ts#L14).

File: packages/stimulus/src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (amiel) [reply packages/stimulus/src/index.ts]: ok, I think I've addressed this and your other comment in 1a94c196cb33274238a221f130a324aadc5c55fa.

**Review by unflxw — APPROVED** [review]
Sorry I left this hanging last year. Fantastic work @amiel, thank you so much!

**Comment by amiel** [comment]
Hi @unflxw. Thank you so much for releasing this. We are now using it in production 🎉 

What is the best way to request stroopwafels?

**Comment by unflxw** [comment]
@amiel Please email support@appsignal.com with a link to this thread! 🍪
